File: src/Base/Sylvester.Api/Logger.fs
namespace Sylvester

open System

/// Logging interface
[<AbstractClass>]
type Logger() =
    static member val IsConfigured = false with get,set
    
    abstract member Info : messageTemplate:string * [<ParamArray>]args:obj[] -> unit
    abstract member Debug : messageTemplate:string * [<ParamArray>]args:obj[] -> unit
    abstract member Error : messageTemplate:string * [<ParamArray>]args:obj[] -> unit
    abstract member Error : ex:Exception * messageTemplate:string * [<ParamArray>]args:obj[] -> unit
    abstract member Begin : messageTemplate:string * [<ParamArray>]args:obj[] -> IDisposable


/// Basic console logger available anywhere
type ConsoleLogger() =
    inherit Logger()

    override __.Debug (messageTemplate:string, [<ParamArray>] args: obj[]) = 
        String.Format(messageTemplate, args) |> printfn "%s"
    override __.Info (messageTemplate:string, [<ParamArray>] args: obj[])  = 
        String.Format(messageTemplate, args) |> printfn "%s"
    override __.Error (messageTemplate:string, [<ParamArray>] args: obj[])  = 
        String.Format(messageTemplate, args) |> printfn "%s"
    override __.Error (ex: Exception, messageTemplate:string, [<ParamArray>] args: obj[])  = 
        String.Format(messageTemplate, args) |> printfn "%s"

    override __.Begin (messageTemplate:string, [<ParamArray>] args: obj[]) = new ConsoleLoggerOperation(messageTemplate, args) :> IDisposable

and ConsoleLoggerOperation(messageTemplate:string, [<ParamArray>] args: obj[]) = 
    let message = String.Format(messageTemplate, args)
    let start = DateTime.Now
    do printfn "Begin %s..." message
    
    interface IDisposable with
        member __.Dispose() = let completed = DateTime.Now.Subtract(start) in do printfn "Completed %s in %.0f ms." message (completed.TotalMilliseconds)

File: src/Base/Sylvester.Api/Logger.cs
using System;

namespace Sylvester
{
    public abstract class Logger
    {
        public abstract class Op : IDisposable
        {
            public Op(Logger l)
            {
                L = l;
            }

            public Logger L { get; }

            protected bool isCompleted = false;

            protected bool isCancelled = false;

            public abstract void Complete();

            public abstract void Cancel();

            public abstract void Dispose();
        }

        public bool IsConfigured { get; protected set; } = false;

        public abstract void Info(string messageTemplate, params object[] args);

        public abstract void Debug(string messageTemplate, params object[] args);

        public abstract void Error(string messageTemplate, params object[] args);

        public abstract void Error(Exception ex, string messageTemplate, params object[] args);

        public abstract Op Begin(string messageTemplate, params object[] args);
    }

    public class ConsoleOp: Logger.Op
    {
        public ConsoleOp(ConsoleLogger l) : base(l) { }


        public override void Complete()
        {
            L.Info("Complete.");
            isCompleted = true;
        }

        public override void Cancel()
        {
            isCancelled = true;
            L.Error("Cancelled");
        }

        public override void Dispose()
        {
            if(!(isCompleted || isCancelled))
            {
                L.Error("Cancelled.");
            }
        }
    }
        
    public class ConsoleLogger : Logger
    {
        public override void Info(string messageTemplate, params object[] args) => Console.WriteLine(messageTemplate, args);

        public override void Debug(string messageTemplate, params object[] args) => Console.WriteLine(messageTemplate, args);

        public override void Error(string messageTemplate, params object[] args) => Console.WriteLine(messageTemplate, args);

        public override void Error(Exception ex, string messageTemplate, params object[] args) => Console.WriteLine(messageTemplate, args);

        public override Op Begin(string messageTemplate, params object[] args) => new ConsoleOp(this);


    }
}
